projects/bootloader/src/bootloader_crc32.c

```diff
@@ -0,0 +1,34 @@
+#include "bootloader_crc32.h"
+
+#include "stdio.h"
+
+#include "bootloader_mcu.h"
+#include "crc32.h"
+#include "flash.h"
+
+// calculates the crc32 for the full address, with 2048 byte increments
+// at a time, returning the final crc32 code
+uint32_t calculated_application_crc32(uintptr_t address, size_t size) {
+  // code array for holding crc32 codes per 2048 bytes
+  // size of array is 54 since
+  // BOOTLOADER_APPLICATION_SIZE / 2048 = 54
+  uint32_t crc32_codes[54];
```

Comment:
54 shouldn't be hardcoded here - instead, use a #define for the length of the buffer (say `BUFFER_LEN`) and then you can declare `crc32_codes` as
```c
uint32_t crc32_codes[BOOTLOADER_APPLICATION_SIZE / BUFFER_LEN + 1];`
```
(with the +1 for safety in case it doesn't align exactly).

You should also check that `size / BUFFER_LEN <= SIZEOF_ARRAY(crc32_codes)` (`SIZEOF_ARRAY` is defined in misc.h and gives the length of the array), because otherwise you'll overflow the end of the array.